### ISSUE TITLE
Newer zip.js needs babel transforms for old browsers

### ISSUE THREAD

**Issue by tophf** [violentmonkey/violentmonkey#1586]
When I reported that the new zip.js uses globalThis which is unsupported in Chrome 55-70, the author fixed it but said it's up to us to maintain compatibility with the old browsers.

There are more problems like destructuring in objects (Chrome 60+) and async generators (Chrome 63+). I guess we'll have to babelize it ourselves, which was something that a reviewer on AMO disallowed, but it might be a personal mistake because they don't object to our custom build of codemirror.js.

Currently, as a temporary workaround, I've pinned an old zip.js in package.json.

**Comment by cyfung1031** [comment]
change to jszip?

https://github.com/Stuk/jszip - 9.1k Star
https://github.com/gildas-lormeau/zip.js - 3k Star

**Comment by tophf** [comment]
It doesn't support workers, which means it'll be slow on a device like smartphone. Anyway, this problem isn't critical for now as the old version works just fine.

**Comment by cyfung1031** [comment]
@tophf  please be noted that jszip can zip more than 65,535 files but fflate cannot.

[jszip-limit-test.html](https://github.com/user-attachments/files/28417783/jszip-limit-test.html)
[fflate-zipsync-limit-test.html](https://github.com/user-attachments/files/28417802/fflate-zipsync-limit-test.html)

See https://github.com/101arrowz/fflate/issues/272

**Comment by tophf** [comment]
Good to know that fflate has sane limits.

**Comment by Rikk** [comment]
Hi @tophf. I think this change of library has caused a loss of timestamp precision (subseconds) in exported files inside the zip:

<img width="637" height="453" alt="Image" src="https://github.com/user-attachments/assets/8d79b469-39d8-44fd-9830-e2494934c32e" />

Also I see a very very small reduction in compression of each file (only a few bytes), insignificant.

**Comment by cyfung1031** [comment]
I have checked the current JavaScript libraries for zip creation... none of them satisfied me.
So I decided to make a new one.
It is almost finish... just to update some doc and benchmark...
I will soon publish it to npm. (size & speed outperform both jszip and fflate)

Yes. `fflate` does not store the extended timestamp. this information is per file entry so the total zip file size will be increased.
both `uzip.js` and `fflate` did not take it into account.

the DOS timestamp just store local time and always even second.

---

<img width="1138" height="498" alt="Image" src="https://github.com/user-attachments/assets/cceb2ad6-b4d5-4a16-997a-d7b21f8e8be9" />

FYI. `zip.js` has included the entire extended metadata. just its implementation is very slow.

**Comment by tophf** [comment]
@Rikk, yeah definitely, but why is this important? If I'm convinced I can try patching fflate the way I already did to expose the time when importing, until cyfung1031's library is published.

**Comment by cyfung1031** [comment]
@tophf  might I ask what is the minimum browser requirement for Violentmonkey (chrome version, firefox version, etc)? I might need to add polyfill if you want to support very old browsers.

**Comment by tophf** [comment]
Violentmonkey currently supports Chrome 61 and Firefox 58, however these are really ancient and not practically usable in the modern times. In Stylus for MV2 I've bumped them to 86/68 after checking on caniuse.com that they support 99% of what modern web does. Someone also said that Chrome around 87 is still being used for something like Windows XP? Anyway, hopefully Violentmonkey will migrate to MV3 and the minimum will be something modern like 128 similarly to Stylus.

**Comment by Rikk** [comment]
> [@Rikk](https://github.com/Rikk), yeah definitely, but why is this important?

It is useful for backup of specific scripts instead of all scripts together. It was a nice detail only our extension had. If the database stores timestamps up to milliseconds, having them mirrored in physical files is good for consistency and reliability. (In a hypothetical case, this could be used to import scripts with precise modification times, even if the config file is missing).

**Comment by tophf** [comment]
Sorry, I still don't understand what's good about milliseconds.

**Comment by tophf** [comment]
OTOH it might make sense if Violentmonkey implements the creation date (currently it doesn't), in which case we can export all three dates:
* modified/updated
* accessed (site visited)
* created
...and add a sort order by created date in the dashboard.

**Comment by cyfung1031** [comment]
> OTOH it might make sense if Violentmonkey implements the creation date (currently it doesn't), in which case we can export all three dates:
> 
> * modified/updated
> * accessed (site visited)
> * created
>   ...and add a sort order by created date in the dashboard.

yes. we are doing these in ScriptCat.

btw, the polyfill etc all done... ~just fixing the documentation lol.~ need to trim down the polyfill size.. wait
(one for 61/58 and one for 86/68 ... up to you)
there is option to keep modifiedAt, lastAccess, createdAt, although your implementation before just added "modifiedAt" to zip.js

will publish within 1~2 days to npm.

**Comment by cyfung1031** [comment]
> [@Rikk](https://github.com/Rikk), yeah definitely, but why is this important? If I'm convinced I can try patching fflate the way I already did to expose the time when importing, until cyfung1031's library is published.

the library is released.

For cr61ff58, i will make a update to fix issue for globalThis and async generator compatibility issue in few days. 
There is `globalThis` in the script which needs Chrome 71+ and Firefox 65+, and `async function*` which needs Chrome 63+ and Firefox 57+.

Remarks: I found that web worker would not boost the speed and there is copy of memory or use of shared memory across process. The way I implemented in JSZipp is a true non UI blocking and faster than fflate - it does iterative compression in each file addition instead of a bulk compression process.

## JSZipp

https://www.npmjs.com/package/web-jszipp
https://github.com/jszipp/jszipp
https://jszipp.github.io/JSZipp/pages/

### Tests / Demo：

https://jszipp.github.io/JSZipp/demo/library-benchmark.html
https://jszipp.github.io/JSZipp/demo/compress.html
https://jszipp.github.io/JSZipp/demo/

---


<img width="857" height="87" alt="Image" src="https://github.com/user-attachments/assets/2b0f08bd-8ce6-44aa-b110-52a7f2dcad18" />

<img width="1121" height="110" alt="Image" src="https://github.com/user-attachments/assets/06bf518f-23ac-4825-9d56-fe196b6b2d8b" />


Change `from "web-jszipp"` to `from "web-jszipp/browser-legacy/cr61ff58"` or `from "web-jszipp/browser-legacy/cr86ff68"` for the version with polyfill (same API interfaces)

for the date, it is by default to do the unix way, which is DOS local time + UTC time for "modifiedAt"

if you want the NTFS way which is modifiedAt + createdAt + lastAccess, change the option to `Dos | Unix | Ntfs` which is the way `zip.js` did.

### https://jszipp.github.io/JSZipp/demo/library-benchmark.html

<img width="393" height="328" alt="Image" src="https://github.com/user-attachments/assets/562096a0-10c4-4981-a724-5d6132c04e2c" />

<img width="388" height="435" alt="Image" src="https://github.com/user-attachments/assets/d6596993-df36-476b-84a9-61eb0c594a2b" />

### demo https://jszipp.github.io/JSZipp/demo/compress.html 

<img width="460" height="135" alt="Image" src="https://github.com/user-attachments/assets/63261d81-16c1-4bd9-b4f3-a9082933fcfc" />

For JSZipp, it is not the same compression payload as `zip.js` which is intended.

---

Details for the timestamp mode option in JSZipp

<img width="927" height="369" alt="Image" src="https://github.com/user-attachments/assets/24b6eabc-393e-44d2-9647-b1a9ccff8d9a" />

**Comment by tophf** [comment]
This was fast, I haven't even looked into adding dates to fflate yet! I'll probably wait for it to be battle-tested by you and others a bit though before switching to it.

**Comment by cyfung1031** [comment]
> This was fast, I haven't even looked into adding dates to fflate yet! I'll probably wait for it to be battle-tested by you and others a bit though before switching to it.

Unlike other libraries, this one you can choose the timestamp format, but i think unix OS will just ignore the createdAt/lastAccess.
so you can still switch to it first and then consider whether VM should make the createdAt/lastAccess dates or not.

For your comparison, ScriptCat stored the creation date and update date - so backup/restore wont lose these info which I believe they are important. 

@Rikk you can use the above demo (compress.html) to test the compression if you are interested. 
I have only tested it in MacOS.

**Comment by tophf** [comment]
Violentmonkey and Tampermonkey have only been storing the last updated/modified date and the primary source of truth is a dedicated property in the exported script object, not the file date, which is set only for user convenience. I don't remember ever seeing a suggestion for a creation date, it's an idea I just had in this thread.

**Comment by Rikk** [comment]
I like the idea of recording creation date, however I don't think site access date would be useful for a script manager (site access times belong to the browser's history imo — and if the user clears the history, the extension would also need to handle that, and private modes).
Given that creation date was never recorded before, the initial value could be the same as the modification date, and if the user imports an older backup, the creation date would be replaced with the older.
Files' timestamp could be a secondary source of truth other than the exported script object.

**Comment by tophf** [comment]
Site access time is not about sites but about a userscript running per se, so there's no connection to browsing history, it's just a convenient metric to detect unused userscripts.

The file's timestamp is used in case the imported script object is incomplete or nonstandard, I'm not aware of any real examples.

**Comment by Rikk** [comment]
I see... Well, in my case, I create scripts for specific sites, therefore access time would identify which sites were visited. Ok, the "detect unused userscripts" case is valid, just not something I need - I'm used to disabling scripts I don't use; maybe others will like it :)

**Comment by cyfung1031** [comment]
> support workers inside extensions without changing their CSP, which we don't want to do, so it'll be slow on a device like smartphone. 

JSZipp@1.0.7: Added plugin option for compression in web worker.

https://jszipp.github.io/JSZipp/demo/worker-compression-demo.html
